File: src/api/appearance.ts
import { ipc, ipcUnit, mocked, shouldMock } from './shared';

/** Where a single custom picture goes. */
export type ImageSlot = 'wallpaper';

/** One of the user's own mascot pictures. */
export interface CustomMascot {
  readonly id: string;
  /** A `data:` URL. */
  readonly url: string;
}

/** Opens a file picker for a picture; null when cancelled. */
export async function pickImage(): Promise<string | null> {
  const { open } = await import('@tauri-apps/plugin-dialog');
  const picked = await open({
    multiple: false,
    filters: [{ name: 'Изображение', extensions: ['png', 'jpg', 'jpeg', 'gif', 'webp'] }],
  });
  return typeof picked === 'string' ? picked : null;
}

/** Copies the picture into the launcher's data folder; returns a data URL. */
export function setAppearanceImage(slot: ImageSlot, path: string): Promise<string> {
  if (shouldMock()) return Promise.reject(new Error('Выбор файла доступен только в приложении'));
  return ipc<string>('set_appearance_image', { slot, path });
}

export function appearanceImage(slot: ImageSlot): Promise<string | null> {
  if (shouldMock()) return mocked<string | null>(null, 50);
  return ipc<string | null>('appearance_image', { slot });
}

export function clearAppearanceImage(slot: ImageSlot): Promise<void> {
  if (shouldMock()) return mocked(undefined, 50);
  return ipcUnit('clear_appearance_image', { slot });
}

export function listCustomMascots(): Promise<CustomMascot[]> {
  if (shouldMock()) return mocked<CustomMascot[]>([], 50);
  return ipc<CustomMascot[]>('list_custom_mascots');
}

/** Copies a picture into the collection. */
export function addCustomMascot(path: string): Promise<CustomMascot> {
  if (shouldMock()) return Promise.reject(new Error('Выбор файла доступен только в приложении'));
  return ipc<CustomMascot>('add_custom_mascot', { path });
}

export function removeCustomMascot(id: string): Promise<void> {
  if (shouldMock()) return mocked(undefined, 50);
  return ipcUnit('remove_custom_mascot', { id });
}

/** The account's skin drawn full-length; null without a skin (offline). */
export function accountFigure(accountId: string): Promise<string | null> {
  if (shouldMock()) return mocked<string | null>(null, 50);
  return ipc<string | null>('account_figure', { accountId });
}

File: src-tauri/src/commands/appearance.rs
use std::path::PathBuf;

use tauri::State;

use crate::accounts::AccountStore;
use crate::appearance::{self, CustomMascot, ImageSlot};
use crate::auth::skin;
use crate::error::Result;
use crate::state::AppState;

/// Copies a picked picture into the slot; returns it ready to display.
#[tauri::command]
pub async fn set_appearance_image(
    state: State<'_, AppState>,
    slot: ImageSlot,
    path: PathBuf,
) -> Result<String> {
    appearance::set_image(&state.paths, slot, &path).await
}

#[tauri::command]
pub async fn appearance_image(state: State<'_, AppState>, slot: ImageSlot) -> Result<Option<String>> {
    appearance::load_image(&state.paths, slot).await
}

#[tauri::command]
pub async fn clear_appearance_image(state: State<'_, AppState>, slot: ImageSlot) -> Result<()> {
    appearance::clear_image(&state.paths, slot).await;
    Ok(())
}

#[tauri::command]
pub async fn list_custom_mascots(state: State<'_, AppState>) -> Result<Vec<CustomMascot>> {
    appearance::list_mascots(&state.paths).await
}

#[tauri::command]
pub async fn add_custom_mascot(state: State<'_, AppState>, path: PathBuf) -> Result<CustomMascot> {
    appearance::add_mascot(&state.paths, &path).await
}

#[tauri::command]
pub async fn remove_custom_mascot(state: State<'_, AppState>, id: String) -> Result<()> {
    appearance::remove_mascot(&state.paths, &id).await
}

/// The account's skin drawn full-length, for the corner mascot. `None` when
/// the account has no skin (offline) or it could not be fetched.
#[tauri::command]
pub async fn account_figure(state: State<'_, AppState>, account_id: String) -> Result<Option<String>> {
    let store = AccountStore::load(&state.paths).await?;
    let Some(url) = store.find(&account_id).and_then(|account| account.skin_url.clone()) else {
        return Ok(None);
    };
    Ok(skin::fetch_body(&state.client(), &url).await)
}
